Add InflightRequest load balancing policy

Gor027 opened the pull request:
The new load balancing policy will consider inflight request metrics for each node/shard for creating a query plan. It will first return replicas with the target shard having the least amount of inflight requests. It will round robin over the replicas with equal inflight requests. The policy considers two replicas to be equal if the target shards of both replicas have inflight requests that may differ by at most 32.

Gor027 committed:
Add InflightRequest load balancing policy

The new load balancing policy will consider inflight request metrics
for each node/shard for creating a query plan. It will first return
replicas with the target shard having the least amount of inflight requests.
It will round robin over the replicas with equal inflight requests. The
policy considers two replicas to be equal if the target shards of both
replicas have inflight requests that may differ by at most 32.

avelanarius commented:
Probably the best way to implement this would be to implement the ["The Power of Two Random Choices" algorithm](https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/) - this is what Java Driver 4.x does (but it only tracks whole-node inflight request numbers, not per shard). In Java Driver 3.x the implementation will be a bit trickier - probably I'll have to duplicate the `RoundRobinLoadBalancingPolicy` and `DCAwareRoundRobinLoadBalancingPolicy` classes.

I'll work on it, while also doing the Java Driver 4.x part (tracking of per-shard inflight numbers to use instead of whole-node numbers).

Gor027 committed:
Add RandomTwoChoice load balancing policy

The policy will first shuffle all replicas using Fisher-Yates shuffle
algorithm, then it will compare first two replicas by inflight requests
and the one with less inflight requests will appear the first in the
query plan.

Gor027 commented:
This was a draft PR and a more general and efficient load-balancing policy will be added soon instead of the `InflightRequest` policy. So, this PR can be closed now.
